packages/nitro-protocol/contracts/MultiAssetHolder.sol

```diff
@@ -2,104 +2,232 @@
      */
     function transfer(
+        uint256 assetIndex, // TODO consider a uint48?
         bytes32 fromChannelId,
-        bytes calldata allocationBytes,
+        bytes memory outcomeBytes,
+        bytes32 stateHash,
+        address challengerAddress,
         uint256[] memory indices
     ) external override {
         // checks
         _requireIncreasingIndices(indices);
-        _requireCorrectAllocationHash(fromChannelId, allocationBytes);
+        _requireChannelFinalized(fromChannelId);
+        _requireMatchingFingerprint(
+            stateHash,
+            challengerAddress,
+            keccak256(outcomeBytes),
+            fromChannelId
+        );
+
+        Outcome.OutcomeItem[] memory outcome = abi.decode(outcomeBytes, (Outcome.OutcomeItem[]));
+        Outcome.AssetOutcome memory assetOutcome = abi.decode(
+            outcome[assetIndex].assetOutcomeBytes,
+            (Outcome.AssetOutcome)
+        );
+        require(
+            assetOutcome.assetOutcomeType == Outcome.AssetOutcomeType.Allocation,
+            '!allocation'
+        );
+        Outcome.AllocationItem[] memory allocation = abi.decode(
+            assetOutcome.allocationOrGuaranteeBytes,
+            (Outcome.AllocationItem[])
+        );
+        address asset = outcome[assetIndex].asset;
+
         // effects and interactions
-        _transfer(fromChannelId, allocationBytes, indices);
+        uint256 initialHoldings;
+        (allocation, initialHoldings) = _transfer(asset, fromChannelId, allocation, indices); // update in place to newAllocation
+        outcome[assetIndex].assetOutcomeBytes = abi.encode(
+            Outcome.AssetOutcome(Outcome.AssetOutcomeType.Allocation, abi.encode(allocation))
+        );
+        outcomeBytes = abi.encode(outcome);
+        bytes32 outcomeHash = keccak256(outcomeBytes);
+        _updateFingerprint(fromChannelId, stateHash, challengerAddress, outcomeHash);
+        emit AllocationUpdated(fromChannelId, assetIndex, initialHoldings);
     }
 
     /**
      * @notice Transfers as many funds escrowed against `guarantorChannelId` as can be afforded for a specific destination in the beneficiaries of the __target__ of that channel. Checks against the storage in this contract.
      * @dev Transfers as many funds escrowed against `guarantorChannelId` as can be afforded for a specific destination in the beneficiaries of the __target__ of that channel. Checks against the storage in this contract.
-     * @param guarantorChannelId Unique identifier for a guarantor state channel.
-     * @param guaranteeBytes The abi.encode of Outcome.Guarantee
-     * @param allocationBytes The abi.encode of AssetOutcome.Allocation for the __target__
-     * @param indices Array with each entry denoting the index of a destination (in the target channel) to transfer funds to. Should be in increasing order. An empty array indicates "all".
      */
     function claim(
```

Comment:
The new code blocks are indeed there to prevent stack too deep errors. There may be other ways to prevent that problem, and I would probably look to tackle that alongside improving readability and ensuring checks/effects/interactions.

As for updating the source channel (aka guarantor channel): what update do you have in mind? There are no updates made to this channel on the base branch (i.e. current behaviour on master), other than to clear the storage in each asset holder when the storage for the target is cleared.

When we move to the exit-format, it becomes important to update the "source" channel because it includes an allocation amount.

Reply:
On `master`, there is a security hole caused by replaying an exit-related action!

When analysing the security of the virtual funding protocol, I think there is an unspoken assumption (at least in my mind) that `claim(G, J)` can only be called _once_. I haven't even thought of the implications of calling it multiple times, but it seems like a strange thing to allow. Therefore, I would assume that on `master`, claiming a guarantee would remove that asset outcome from the guarantor channel's outcome.

That said, you can argue that `claim(G, J)` will have no effect, since `G.holdings` will be 0 after calling it the 1st time. (Or, it'll be positive, and calling it the second time should have the same effect.) These are somewhat complex lines of argument, though.

_All that being said_, I don't think there's any issue with leaving it as is, as the implementation will need to change (3) accordingly.

> it becomes important to update the "source" channel because it includes an allocation amount.

It is also very important because the "source" channel has many outcome items in addition to the guarantee being claimed.